Adds the Materials for PWAs explainer (#997)
MicrosoftEdge/MSEdgeExplainers: Materials/explainer.md

[Windows 11](https://learn.microsoft.com/en-us/windows/apps/design/signature-experiences/materials) also has a similar concept, with `Mica`, `Mica Alt` and `Acrylic`.

![Copilot app on Windows 11 displaying a left panel with a translucent material background](winMaterial.png)

This explainer addresses allowing web applications to use materials that might be available in the underlying platform. When enabled, and if the platform supports it, the frame of the installed web application will use the specified visual effect, and if the background of the page is transparent the material would be visible.

> **Note:** A developer can already create and mimic experiences based on opaque surfaces. By using images and paterns there is a high chance that a desired platform aesthetic can be achieved (like _brushed-metal_). With newer and more powerful hardware, platforms are starting to add transparency effects to their UX surfaces. The intention of this API is to allow a web application's frame to use materials with transparency effects like the ones in the screenshots below.

## Goals

- Enable the frame of an installed web application to use OS materials.

> - Design a solution that will scale to future design systems.

Reviewer: ```suggestion
- Design a solution that will scale to future OS materials.
```

(design system has a very specific meaning, which doesn't match what you're trying to say here).

## Non-goals

> - Define or override material usage for the application's UX elements like context menus and tooltips (This is handled by the platform).

Reviewer: ```suggestion
- Define a solution to apply material to UI elements like context menus and tooltips (this is already handled by the platform).
```

> - Support devices that do not have content behind a frame/window that 'bleeds' through. As an example, (non-desktop) mobile devices generally don't have regions of the UX where the user would be able to see content behind the application's frame.

Reviewer: Not sure how to rephrase, but I would maybe suggest something like this:

```suggestion
- Support devices, such as mobile devices, that do not have support for a windowing system where the content that's behind the application frame would be visible if that frame was transparent.
```

## Proposed Solutions

### Manifest enabled Materials

> An application manifest is a JSON document that contains startup parameters and application defaults for when a web application is launched. Both options presented below can scale by adding new values for the manifest key.

Reviewer: ```suggestion
A web application manifest is a JSON file that contains startup parameters and application defaults for when a web application is launched and installed. We are presenting two different options, both based on adding a new manifest member.
```

#### Option 1: Material Abstraction

> We propose adding a new `base_material` member to the manifest file, which hints to the UA that the frame should be of a certain type if supported. The `base_material` key can have a value of:

Reviewer: ```suggestion
We propose adding a new `base_material` member to the manifest file, which hints to the UA that the frame should be of a certain type, if the operating sustem supports it. The `base_material` member can be set to the following values:
```

-  `opaque`: the frame of the installed web app does not have any material applied to it. This is the default value.
- `translucent`: the frame of the installed web app is of a translucent material. This is a softened effect that prevents full visibility.

> - `transparent`: the frame of the app has a transparent-like effect. This material allows the background to pass through with less significant scattering, providing a clearer, less distorted view than `translucent`.

Reviewer: Yeah, transparent means that the objects behind can clearly be seen (no scattering of light at all).
I sounds like both non-opaque values here are really about different levels of light scattering. They're both translucent, but one scatters more than the other.

> - `transparent`: the frame of the app has a transparent-like effect. This material allows the background to pass through with less significant scattering, providing a clearer, less distorted view than `translucent`.

Author: We need to come up with better names 😅.

> **NOTE**:
> The way a UA maps the values to the available platform materials is completely up to the implementation. Different platforms have different number of materials and this explainer aims to provide an abstraction via an offering of a limited set of materials available for web apps.

```JSON
{
  "name": "Awesome Web App",
  "short_name": "Awesome App",
  "id": "aweapp",
  ...
  "start_url": "/index.html",
  "display": "standalone",
  "base_material": "translucent"
}
```

> The web developer needs to specify the background of the document as transparent (`background-color: transparent`) to remove the color from the viewport and let the web content appear as on top of the translucent frame. Any additional in-app effects, such as materials and tinting can be achieved with CSS like `backdrop-filter`.

Reviewer: ```suggestion
To let the base material be visible in their app, web developers must set the background of the document to be transparent (`background-color: transparent`). This removes the color from the viewport and lets the web content appear on top of the translucent frame. Any additional in-app effects, such as materials and tinting can be achieved with CSS, for example by using `backdrop-filter`.
```

#### Option 2: Explicit Material Platform Values

> We propose adding a property to a PWA manifest that allows an app to declare a prioritized list of materials that it would like to use as its background if possible, conceptually similar to `font-family` matching the the phrasing the OS uses for the material:

Reviewer: ```suggestion
We propose introducing the `background_material` member to the web app manifest. This member allows apps to declare a prioritized list of materials that they wish to use as frame background if possible. This is conceptually similar to the `font-family` CSS property, which also goes through a list of prioritized fonts to find one that can be used.

The values of the `background_material` member match how specific materials are named in various operating systems:
```

```JSON
{
  "name": "Awesome Web App",
  "short_name": "Awesome App",
  "id": "aweapp",
  ...
  "start_url": "/index.html",
  "display": "standalone",
  "background_material": ["windows-mica", "macos-ultrathick", "windows-acrylic", "#87ceeb"]
}
```

> The browser would attempt to use the given materials if possible, with the (optional) final color being a fallback.

Reviewer: I don't see a need for the final color value since you're proposing to detect the material at runtime anyway. And having it here means that the values of the array can have two types.

### Material Detection

> To allow the developer to gracefully degrade the experience if a platform does not support materials an additional media query is required. This query would allow the site to conditionally make the content on top of it transparent or give it a specific color:

Reviewer: ```suggestion
To allow developers to set their app's background to transparent only when the platform supports materials, and use a different background color if not, we propose adding a media query:
```

```CSS
/*option 1*/

@media (base-material: 'opaque'){
  body{

>     background-color: aliceblue);

Reviewer 2: ```suggestion
    background-color: aliceblue;
```

  }
}

> 

Reviewer: Agreed, very important to add because devs will always need both at-media rules.

/*option 2*/

> @media (background-material: 'windows-mica'){

Reviewer: I think this should be expanded to show multiple values:

```suggestion
@media (background-material: 'windows-mica'), (background-material: 'macos-ultrathick') {
```

Also, how would a developer know if none of these values match? What would be the media query?

  body{
    background-color: rgba(0, 0, 0, 0));
  }
}
```

## Considered Alternatives

> ### CSS "Color"

Reviewer: ```suggestion
### Using a new <color> CSS data type
```

> We thought about having a special CSS "color" that would represent the material so it would be easily applied to different surfaces in the viewport. The problem is that this more a frame property than a DOM object property. Having the frame be of a certain material allows the developer to create the UX they want. Specifying a CSS background color and assigning this to different areas in the viewport conflicts with the overall UA's default background and the color of the frame provided by the underlying platform.

Reviewer: ```suggestion
We considered introducing a new value to the `<color>` CSS data type, which would represent the material. This value could then be easily applied by developers to the background of the app, or to different surfaces within the app.

The limitation is that the CSS styles defined by developers apply to elements that are within the viewport, and not to the app frame itself. Using a CSS background color for different areas in the viewport would conflict with the UA's default background of the app frame provided by the underlying platform.
```
